src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs

```diff
@@ -94,88 +94,76 @@ protected override void Replicate()
+                foreach (var kvp in batch.Attachments)
                 {
-                    foreach (var kvp in batch.Attachments)
+                    using (stats.For(ReplicationOperation.Outgoing.AttachmentRead))
                     {
-                        using (stats.For(ReplicationOperation.Outgoing.AttachmentRead))
+                        var attachment = kvp.Value;
+                        try
                         {
-                            using (var attachment = kvp.Value)
-                            {
-                                try
-                                {
-                                    attachment.WriteStream(_stream, _tempBuffer);
-                                    stats.RecordAttachmentOutput(attachment.Stream.Length);
-                                }
-                                catch
-                                {
-                                    if (Logger.IsInfoEnabled)
-                                        Logger.Info($"Failed to write Attachment stream {FromToString}");
-
-                                    throw;
-                                }
-                            }
+                            attachment.WriteStream(_stream, _tempBuffer);
+                            stats.RecordAttachmentOutput(attachment.Stream.Length);
+                        }
+                        catch
+                        {
+                            if (Logger.IsInfoEnabled)
+                                Logger.Info($"Failed to write Attachment stream {FromToString}");
+
+                            throw;
                         }
                     }
                 }
+            }
 
-                _stream.Flush();
-                sw.Stop();
+            _stream.Flush();
+            sw.Stop();
 
-                var (type, reply) = HandleServerResponse(getFullResponse: true);
+            var (type, reply) = HandleServerResponse(getFullResponse: true);
 
-                batch.LastAcceptedChangeVector = _lastAcceptedChangeVectorFromShard = reply.DatabaseChangeVector;
-                batch.LastEtagAccepted = _lastSentDocumentEtag = reply.LastEtagAccepted;
+            batch.LastAcceptedChangeVector = _lastAcceptedChangeVectorFromShard = reply.DatabaseChangeVector;
+            batch.LastEtagAccepted = _lastSentDocumentEtag = reply.LastEtagAccepted;
 
-                if (type == ReplicationMessageReply.ReplyType.MissingAttachments)
-                {
-                    MissingAttachmentsInLastBatch = true;
-                    MissingAttachmentMessage = reply?.Exception;
-                }
-            }
-            finally
+            if (type == ReplicationMessageReply.ReplyType.MissingAttachments)
             {
-                foreach (var item in batch.Items)
-                {
-                    item.Dispose();
-                }
+                MissingAttachmentsInLastBatch = true;
+                MissingAttachmentMessage = reply?.Exception;
```

Comment:
where is the `finally` with the `item.Dispose();` ?

Reply:
because we are disposing the batch later in `ShardedIncomingReplicationHandler`